cmd/config/docs/api-conventions/functions-spec.md

```diff
@@ -327,16 +327,22 @@ A function SHOULD preserve comments when input serialization format is YAML.
 This allows for human authoring of configuration to coexist with changes made by
 functions.
 
-### Annotations
-
-The orchestrator annotates resources in the wire format with annotation prefix
-`config.kubernetes.io`. These annotations are not persisted when the
-orchestrator writes the resources to the filesystem. The orchestrator sets this
-annotation when reading files from the local filesystem and removes the
-annotation when writing the output of functions back to the filesystem.
-
-In general, a function MUST NOT modify these annotations except the ones
-explicitly listed below.
+### Control Annotations
+
+For orchestration purposes, the orchestrator will use a set of annotations,
+referred to as _control annotations_, on resources in `Resources.items`. These
+annotations are not persisted to resource manifests on the filesystem: The
+orchestrator sets this annotation when reading files from the local filesystem
+and removes the annotation when writing the output of functions back to the
+filesystem.
```

Comment:
I think there's a nuance here:

1) `Is this guidance actually specific to orchestrators?`

In the spec, the language is that functions "SHOULD NOT" set control annotations, it's not a "MUST NOT". The former explicitly allows for functions (that know what they're doing) to set these annotations:

```
SHOULD NOT   This phrase, or the phrase "NOT RECOMMENDED" mean that
   there may exist valid reasons in particular circumstances when the
   particular behavior is acceptable or even useful, but the full
   implications should be understood and the case carefully weighed
   before implementing any behavior described with this label.
```

Using your example, if these a function that wants to read/write to filesystem (instead of orchestrator doing this), it can do so. Fun fact: one of the first functions ever written was doing exactly that in TS:

https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/demo-functions/src/read_yaml.ts
https://github.com/GoogleContainerTools/kpt-functions-sdk/blob/master/ts/demo-functions/src/write_yaml.ts

In practice, no one uses these functions, since it's a basic functionality that orchestrator provides.

2) `If someone writing a tool that does config IO but does not use functions, is our recommendation around these annotations still valid?`

I would argue that we should just standardize this as part of the functions spec. Otherwise, it's not clear what the interoperability story is for that tool. What other processes does such a non-function tool need to talk to? Do does other tools need to conform to some API, etc.? If you follow this through, you'll end up effectively re-inventing the functions spec.

Reply:
> I would argue that those tools should just conform to the functions spec. Otherwise, it's not clear what the interoperability story is for that tool. What other processes does such a non-function tool need to talk to?

In practice the function orchestrators support receiving a stream of API objects, which they internally wrap in ResourceList. This means that they can take input from any tool that emits the raw APIs, which are arguably the truly universal interchange format for k8s stuff. I don't feel strongly about this, but the idea behind recommending that other tools that read config emit these annotations would be even broader interoperability / standardization. That was the impression I got from the original document. I have no idea if any non-function tools picked up the standard.

Reply:
To close the loop on this: Yes, orchestrators are free to accept input in any format (multi-object YAML, a List, etc.) and convert it to the schema this document is concerned with. This document is only concerned with ResourceList spec.